x-pack/plugins/alerting_builtins/server/alert_types/geo_threshold/geo_threshold.ts

```diff
@@ -0,0 +1,270 @@
+      )
+  );
+}
+
+export const getGeoThresholdExecutor = ({ logger: log }: { logger: Logger }) =>
+  async function ({
+    previousStartedAt: currIntervalStartTime,
+    startedAt: currIntervalEndTime,
+    services,
+    params,
+    alertId,
+    state,
+  }: {
+    previousStartedAt: Date | null;
+    startedAt: Date;
+    services: AlertServices;
+    params: {
+      index: string;
+      indexId: string;
+      geoField: string;
+      entity: string;
+      dateField: string;
+      trackingEvent: string;
+      boundaryType: string;
+      boundaryIndexTitle: string;
+      boundaryIndexId: string;
+      boundaryGeoField: string;
+      boundaryNameField?: string;
+    };
+    alertId: string;
+    state: AlertTypeState;
+  }) {
+    const { shapesFilters, shapesIdsNamesMap } = state.shapesFilters
+      ? state
+      : await getShapesFilters(
+          params.boundaryIndexTitle,
+          params.boundaryGeoField,
+          params.geoField,
+          services.callCluster,
+          log,
+          alertId,
+          params.boundaryNameField
+        );
+
+    const executeEsQuery = await executeEsQueryFactory(params, services, log, shapesFilters);
+
+    // Start collecting data only on the first cycle
+    if (!currIntervalStartTime) {
+      log.info(`alert ${GEO_THRESHOLD_ID}:${alertId} alert initialized. Collecting data`);
+      // Consider making first time window configurable?
+      const tempPreviousEndTime = new Date(currIntervalEndTime);
+      tempPreviousEndTime.setMinutes(tempPreviousEndTime.getMinutes() - 5);
+      const prevToCurrentIntervalResults:
+        | SearchResponse<unknown>
+        | undefined = await executeEsQuery(tempPreviousEndTime, currIntervalEndTime);
+      return {
+        prevLocationArr: transformResults(
+          prevToCurrentIntervalResults,
+          params.dateField,
+          params.geoField
+        ),
+      };
+    }
+
+    const currentIntervalResults: SearchResponse<unknown> | undefined = await executeEsQuery(
+      currIntervalStartTime,
+      currIntervalEndTime
+    );
+    // No need to compare if no changes in current interval
+    if (!_.get(currentIntervalResults, 'hits.total.value')) {
+      return state;
+    }
+
+    const currLocationArr: LatestEntityLocation[] = transformResults(
+      currentIntervalResults,
+      params.dateField,
+      params.geoField
+    );
+
+    const movedEntities: EntityMovementDescriptor[] = getMovedEntities(
+      currLocationArr,
+      state.prevLocationArr,
+      params.trackingEvent
+    );
+
+    // Create alert instances
+    movedEntities.forEach(({ entityName, currLocation, prevLocation }) => {
+      const toBoundaryName = shapesIdsNamesMap[currLocation.shapeId] || '';
+      const fromBoundaryName = shapesIdsNamesMap[prevLocation.shapeId] || '';
+      services
+        .alertInstanceFactory(`${entityName}-${toBoundaryName || currLocation.shapeId}`)
```

Comment:
I'm not sure what `currLocation.shapeId` evaluates to, but you will want the value sent to `alertInstanceFactory()` to be consistent over time for the different instances you are dealing with (each `movedEntity`).  Eg, say `movedEntity` contains one object, for two different invocations of the the alert executor (ie, separated by the alert interval).  For both alert executor calls that entity is logically the same entity - but perhaps it "moved".  In this case, you want the param to `alertInstanceFactory()` to NOT include the location, because that would end up generating a new instance.

It depends on how you want to handle this, but I'd think if this is tracking objects somehow, the alert instance id (the param to `alertInstanceFactory()`), each object should have it's own unique instance id that does not change over time.

Reply:
Also worth noting the instance ids will show up in the alert details page - each instance will appear as row with a bit of status info (how long it's been active, etc), and the instance id should something identifiable to customers.  As an example from o11y, the instance id is often a host name, a service name, maybe combo host-service name, etc (depending on the alert type).

Reply:
> each object should have it's own unique instance id that does not change over time.

Originally we did this. Believe it or not, the way it works currently is intended behavior :grimacing:. This was a topic of heavy discussion along the way with this being the final result. We're aware of the pitfalls you've identified, but ultimately decided it provided the most useful version of the details page for now, acting as a sort of running manifest of points and their crossings into geographic regions. 

I think ultimately this will evolve quite a bit in the near future. For threshold and other geo alert options, we'd be interested in possibly providing the user more information on the current state of tracked points/shapes. I'm not sure if what we're interested in will fit on this details page or perhaps in a separately generated dashboard that's something altogether different. Will definitely keep you and the team looped in thhough!